types/vote_set_test.go

```diff
@@ -472,3 +474,153 @@ func TestMakeCommit(t *testing.T) {
 	}
 
 }
+
+func TestVoteSetInvalid(t *testing.T) {
+	vs := (*VoteSet)(nil)
+	require.Equal(t, vs.Round(), -1, "nil voteSet returns -1")
+	require.Equal(t, vs.Type(), byte(0x00), "nil voteSet returns the zero value type")
+	require.Equal(t, vs.Height(), 0, "nil voteSet returns the zero value height")
+	require.Equal(t, vs.Size(), 0, "nil voteSet returns the zero value size")
+	require.False(t, vs.HasTwoThirdsAny(), "nil voteSet cannot say if it has 2/3")
+	require.False(t, vs.HasTwoThirdsMajority(), "nil voteSet cannot return 2/3 status")
+	require.False(t, vs.IsCommit(), "a nil voteSet cannot tell if it is a commit")
+	require.False(t, vs.HasAll(), "a nil voteSet cannot say if HasAll()")
+	require.Nil(t, vs.MakeCommit(), "a nil voteSet cannot makeCommit")
+	require.Contains(t, vs.StringShort(), "nil-VoteSet")
+	require.Contains(t, vs.String(), "nil-VoteSet")
+	require.Panics(t, func() { vs.SetPeerMaj23("foo", BlockID{}) }, "cannot set majority 2/3 on a nil voteSet")
+	require.Nil(t, vs.GetByAddress([]byte("foo")), "cannot get a vote by address from a nil voteSet")
+	blockID, is := vs.TwoThirdsMajority()
+	require.False(t, is, "cannot be the 2/3 majority if nil")
+	require.Equal(t, blockID, BlockID{}, "expecting the zero-value of BlockID")
+
+	vs = new(VoteSet)
+	require.Panics(t, func() { vs.GetByAddress([]byte("foo")) }, "non-existent address should panic")
+
+	require.Panics(t, func() { NewVoteSet("foo", 0, 1, 0x01, nil) }, "height 0 shouldn't work")
+}
+
+func TestVoteSetEndToEnd(t *testing.T) {
+	vs := (*VoteSet)(nil)
+	require.Nil(t, vs.GetByIndex(0), "should return nil for any index when nil")
+	require.Nil(t, vs.GetByIndex(-1), "should return nil for any index when nil")
+	require.Nil(t, vs.GetByIndex(11), "should return nil for any index when nil")
+
+	require.Panics(t, func() { vs.AddVote(nil) }, "nil voteSet should panic")
+
+	vs, vsl, privValidators := randVoteSet(25, 2, VoteTypePrecommit, 10, 1)
+	require.Equal(t, vs.Size(), 10, "expecting 10 validators as set in")
+	require.Nil(t, vs.GetByIndex(0), "0: hasn't yet cast their vote")
+	require.Nil(t, vs.GetByIndex(1), "1: hasn't yet cast their vote")
+
+	// Invalid/nil/non-dereferenceable vote
+	added, err := vs.AddVote(nil)
+	require.False(t, added, "nil votes can't be dereferenced")
+	require.NotNil(t, err, "nil votes can't be dereferenced")
+
+	// Malicious vote
+	added, err = vs.AddVote(&Vote{
+		ValidatorIndex: -1,
```

Comment:
lets try one with index too high too?

Reply:
Yes we have a high one here https://github.com/tendermint/tendermint/pull/725/files/1b4593fc2945f1af0245c709a973f2a254751851#diff-b751ab127cf8b221ebc8084facc19a38R615

But I can def it clearer here as well.